### ISSUE TITLE
running tests: `catkin test` alias doesn't always work

### ISSUE THREAD

**Issue by k-okada** [catkin/catkin_tools#139]
Hi, I have packages with rostest code which works ok with `catkin run_tests` but fails on `catkin test`, is it correct behavior?
Please see https://gist.github.com/anonymous/f5296c1ae4c5e778c2db

**Comment by wjwwood** [comment]
The `test` target is setup by CTest, as I explained in this ROS answers question (which you asked :stuck_out_tongue:):

http://answers.ros.org/question/136951/how-to-specify-test-packge-in-catkin_make-test/

At any rate the `test` target should work, but I couldn't tell you why it's failing, since you didn't post any error output. Can you please add the output from the failure?

**Comment by wjwwood** [comment]
Sorry I missed the part where you said `catkin test` versus `catkin_make test`. I think the `test` alias (which is what you are invoking) is just incorrect. It would still help to have your error output to confirm that is the issue.

**Comment by wjwwood** [comment]
I'm not sure what the expected behavior is, @jbohren added it here:

https://github.com/catkin/catkin_tools/commit/8803c8964b0ae9805443f8eaaff3ccca176c5a86#diff-954512ee3976ccebcf65cb4af2cbcd3e

Maybe he can comment on what it's supposed to be doing.

**Comment by k-okada** [comment]
`catkin run_tests` works ok but `catkin test` returns error
```
k-okada@kokada-virtual-machine:/tmp/aaa$ catkin test
==> Expanding alias 'test' from 'catkin test' to 'catkin build --verbose --make-args test --'
---------------------------------------------
Profile:                     default
Extending:          [cached] /opt/ros/indigo
Workspace:                   /tmp/aaa
Source Space:       [exists] /tmp/aaa/src
Build Space:        [exists] /tmp/aaa/build
Devel Space:        [exists] /tmp/aaa/devel
Install Space:     [missing] /tmp/aaa/install
DESTDIR:                     None
---------------------------------------------
Isolate Develspaces:         False
Install Packages:            False
Isolate Installs:            False
---------------------------------------------
Additional CMake Args:       None
Additional Make Args:        test
Additional catkin Make Args: None
---------------------------------------------
Workspace configuration appears valid.
--------------------------------------------- 
Found '1' packages in 0.0 seconds. 
Starting ==> testtest                                                                                           
[testtest] ==> '/tmp/aaa/build/testtest/build_env.sh /usr/bin/make cmake_check_build_system' in '/tmp/aaa/build/testtest' 
[testtest] <== '/tmp/aaa/build/testtest/build_env.sh /usr/bin/make cmake_check_build_system' finished with return code '0' 
[testtest] ==> '/tmp/aaa/build/testtest/build_env.sh /usr/bin/make test -j1 -l1' in '/tmp/aaa/build/testtest'   
Running tests...                                                                                                
Test project /tmp/aaa/build/testtest                                                                            
    Start 1: _ctest_testtest_rostest_test.test                                                                  
1/1 Test #1: _ctest_testtest_rostest_test.test ...***Failed    2.22 sec                                         
0% tests passed, 1 tests failed out of 1                                                                        
Total Test time (real) =   2.22 sec                                                                             
The following tests FAILED:                                                                                     
[build -  1 - _ctest_testtest_rostest_test.test (Failed)                                                               
Errors while running CTest                                                                                      
make: *** [test] Error 8                                                                                        
[testtest] <== '/tmp/aaa/build/testtest/build_env.sh /usr/bin/make test -j1 -l1' failed with return code '2'    
[build] There were errors:                                                                                      

    Failed to build package 'testtest' because the following command:

        # Command run in directory: /tmp/aaa/build/testtest
        /tmp/aaa/build/testtest/build_env.sh /usr/bin/make test -j1 -l1

    Exited with return code: 2 
[build] Runtime: 2.8 seconds 
k-okada@kokada-virtual-machine:/tmp/aaa$ catkin run_tests
==> Expanding alias 'run_tests' from 'catkin run_tests' to 'catkin build --verbose --catkin-make-args run_tests --'
---------------------------------------------
Profile:                     default
Extending:          [cached] /opt/ros/indigo
Workspace:                   /tmp/aaa
Source Space:       [exists] /tmp/aaa/src
Build Space:        [exists] /tmp/aaa/build
Devel Space:        [exists] /tmp/aaa/devel
Install Space:     [missing] /tmp/aaa/install
DESTDIR:                     None
---------------------------------------------
Isolate Develspaces:         False
Install Packages:            False
Isolate Installs:            False
---------------------------------------------
Additional CMake Args:       None
Additional Make Args:        None
Additional catkin Make Args: run_tests
---------------------------------------------
Workspace configuration appears valid.
--------------------------------------------- 
Found '1' packages in 0.0 seconds. 
Starting ==> testtest                                                                                           
[testtest] ==> '/tmp/aaa/build/testtest/build_env.sh /usr/bin/make cmake_check_build_system' in '/tmp/aaa/build/testtest' 
[testtest] <== '/tmp/aaa/build/testtest/build_env.sh /usr/bin/make cmake_check_build_system' finished with return code '0' 
[testtest] ==> '/tmp/aaa/build/testtest/build_env.sh /usr/bin/make run_tests -j1 -l1' in '/tmp/aaa/build/testtest' 
Built target tests                                                                                              
Built target clean_test_results                                                                                 
-- run_tests.py: execute commands                                                                               
  /opt/ros/indigo/share/rostest/cmake/../../../bin/rostest --pkgdir=/tmp/aaa/src/testtest --package=testtest --results-filename test.xml /tmp/aaa/src/testtest/test.test 
... logging to /home/k-okada/.ros/log/rostest-kokada-virtual-machine-25160.log                                  
[ROSUNIT] Outputting test results to /tmp/aaa/build/testtest/test_results/testtest/rostest-test.xml             
testtesttest ... ok                                                                                             
                                                                                                                
[ROSTEST]-----------------------------------------------------------------------                                
                                                                                                                
[testtest.rosunit-testtest/test_one_equals_one][passed]                                                         
                                                                                                                
SUMMARY                                                                                                         
 * RESULT: SUCCESS                                                                                              
 * TESTS: 1                                                                                                     
 * ERRORS: 0                                                                                                    
 * FAILURES: 0                                                                                                  
                                                                                                                
rostest log file is in /home/k-okada/.ros/log/rostest-kokada-virtual-machine-25160.log                          
-- run_tests.py: verify result "/tmp/aaa/build/testtest/test_results/testtest/rostest-test.xml"                 
Built target _run_tests_testtest_rostest_test.test                                                              
Built target _run_tests_testtest_rostest                                                                        
Built target _run_tests_testtest                                                                                
Built target run_tests                                                                                          
[testtest] <== '/tmp/aaa/build/testtest/build_env.sh /usr/bin/make run_tests -j1 -l1' finished with return code '0' 
Finished <== testtest [ 3.2 seconds ]                                                                           
[build] Finished.                                                                                               
[build] Runtime: 3.2 seconds 
k-okada@kokada-virtual-machine:/tmp/aaa$ 
```

**Comment by jbohren** [comment]
As @wjwwood said, `catkin test` is for the CTest `test` target, whereas `catkin run_tests` is for the rostest `run_tests` target.

**Comment by wjwwood** [comment]
@k-okada Does this fail with `catkin_make test` or `catkin_make_isolated --catkin-make-args test`?

**Comment by k-okada** [comment]
both worked fine
```
k-okada@kokada-virtual-machine:/tmp/aaa$ catkin_make test
Base path: /tmp/aaa
Source space: /tmp/aaa/src
Build space: /tmp/aaa/build
Devel space: /tmp/aaa/devel
Install space: /tmp/aaa/install
Creating symlink "/tmp/aaa/src/CMakeLists.txt" pointing to "/opt/ros/indigo/share/catkin/cmake/toplevel.cmake"
####
#### Running command: "cmake /tmp/aaa/src -DCATKIN_DEVEL_PREFIX=/tmp/aaa/devel -DCMAKE_INSTALL_PREFIX=/tmp/aaa/install -G Unix Makefiles" in "/tmp/aaa/build"
####
-- The C compiler identification is GNU 4.8.2
-- The CXX compiler identification is GNU 4.8.2
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Using CATKIN_DEVEL_PREFIX: /tmp/aaa/devel
-- Using CMAKE_PREFIX_PATH: /opt/ros/indigo
-- This workspace overlays: /opt/ros/indigo
-- Found PythonInterp: /usr/bin/python (found version "2.7.6") 
-- Using PYTHON_EXECUTABLE: /usr/bin/python
-- Using Debian Python package layout
-- Using empy: /usr/bin/empy
-- Using CATKIN_ENABLE_TESTING: ON
-- Call enable_testing()
-- Using CATKIN_TEST_RESULTS_DIR: /tmp/aaa/build/test_results
-- Looking for include file pthread.h
-- Looking for include file pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found gtest sources under '/usr/src/gtest': gtests will be built
-- Using Python nosetests: /usr/bin/nosetests-2.7
-- catkin 0.6.11
-- BUILD_SHARED_LIBS is on
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- ~~  traversing 1 packages in topological order:
-- ~~  - testtest
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- +++ processing catkin package: 'testtest'
-- ==> add_subdirectory(testtest)
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/aaa/build
####
#### Running command: "make test -j1 -l1" in "/tmp/aaa/build"
####
Running tests...
Test project /tmp/aaa/build
    Start 1: _ctest_testtest_rostest_test.test
1/1 Test #1: _ctest_testtest_rostest_test.test ...   Passed    1.69 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   1.69 sec
k-okada@kokada-virtual-machine:/tmp/aaa$ catkin clean -all
usage: catkin clean [-h] [--workspace WORKSPACE] [--profile PROFILE] [-a] [-b]
                    [-d] [-i] [-c] [-s] [-o]
catkin clean: error: argument -a/--all: ignored explicit argument 'll'
k-okada@kokada-virtual-machine:/tmp/aaa$ catkin clean -a
[clean] Removing buildspace: /tmp/aaa/build
[clean] Removing develspace: /tmp/aaa/devel
k-okada@kokada-virtual-machine:/tmp/aaa$ catkin_make_isolated --catkin-make-args test
Base path: /tmp/aaa
Source space: /tmp/aaa/src
Build space: /tmp/aaa/build_isolated
Devel space: /tmp/aaa/devel_isolated
Install space: /tmp/aaa/install_isolated
Additional make Arguments for catkin packages: test
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~  traversing 1 packages in topological order:
~~  - testtest
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The packages or cmake arguments have changed, forcing cmake invocation

==> Processing catkin package: 'testtest'
==> Creating build directory: 'build_isolated/testtest'
==> cmake /tmp/aaa/src/testtest -DCATKIN_DEVEL_PREFIX=/tmp/aaa/devel_isolated/testtest -DCMAKE_INSTALL_PREFIX=/tmp/aaa/install_isolated -G Unix Makefiles in '/tmp/aaa/build_isolated/testtest'
-- The C compiler identification is GNU 4.8.2
-- The CXX compiler identification is GNU 4.8.2
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Using CATKIN_DEVEL_PREFIX: /tmp/aaa/devel_isolated/testtest
-- Using CMAKE_PREFIX_PATH: /opt/ros/indigo
-- This workspace overlays: /opt/ros/indigo
-- Found PythonInterp: /usr/bin/python (found version "2.7.6") 
-- Using PYTHON_EXECUTABLE: /usr/bin/python
-- Using Debian Python package layout
-- Using empy: /usr/bin/empy
-- Using CATKIN_ENABLE_TESTING: ON
-- Call enable_testing()
-- Using CATKIN_TEST_RESULTS_DIR: /tmp/aaa/build_isolated/testtest/test_results
-- Looking for include file pthread.h
-- Looking for include file pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found gtest sources under '/usr/src/gtest': gtests will be built
-- Using Python nosetests: /usr/bin/nosetests-2.7
-- catkin 0.6.11
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/aaa/build_isolated/testtest
==> make test -j1 -l1 in '/tmp/aaa/build_isolated/testtest'
Running tests...
Test project /tmp/aaa/build_isolated/testtest
    Start 1: _ctest_testtest_rostest_test.test
1/1 Test #1: _ctest_testtest_rostest_test.test ...   Passed    1.50 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   1.50 sec
<== Finished processing package [1 of 1]: 'testtest'
```

**Comment by wjwwood** [comment]
I cannot reproduce this, I get this:

- `catkin test`
```
==> Expanding alias 'test' from 'catkin test' to 'catkin build --verbose --make-args test --'
Creating build space directory, '/private/tmp/catkin_test_target_ws/build'
-----------------------------------------------------------------------
Profile:                     default
Extending:             [env] /Users/william/indigo/install
Workspace:                   /private/tmp/catkin_test_target_ws
Source Space:       [exists] /private/tmp/catkin_test_target_ws/src
Build Space:        [exists] /private/tmp/catkin_test_target_ws/build
Devel Space:       [missing] /private/tmp/catkin_test_target_ws/devel
Install Space:     [missing] /private/tmp/catkin_test_target_ws/install
DESTDIR:                     None
-----------------------------------------------------------------------
Isolate Develspaces:         False
Install Packages:            False
Isolate Installs:            False
-----------------------------------------------------------------------
Additional CMake Args:       None
Additional Make Args:        test
Additional catkin Make Args: None
-----------------------------------------------------------------------
Workspace configuration appears valid.
-----------------------------------------------------------------------
Found '1' packages in 0.0 seconds.
Starting ==> testtest

[testtest] ==> '/private/tmp/catkin_test_target_ws/build/testtest/build_env.sh /usr/local/bin/cmake /private/tmp/catkin_test_target_ws/src/testtest -DCATKIN_DEVEL_PREFIX=/private/tmp/catkin_test_target_ws/devel -DCMAKE_INSTALL_PREFIX=/private/tmp/catkin_test_target_ws/install' in '/private/tmp/catkin_test_target_ws/build/testtest'
-- The C compiler identification is AppleClang 6.0.0.6000056
-- The CXX compiler identification is AppleClang 6.0.0.6000056
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Using CATKIN_DEVEL_PREFIX: /tmp/catkin_test_target_ws/devel
-- Using CMAKE_PREFIX_PATH: /Users/william/indigo/install
-- This workspace overlays: /Users/william/indigo/install
-- Found PythonInterp: /usr/local/bin/python (found version "2.7.9")
-- Using PYTHON_EXECUTABLE: /usr/local/bin/python
-- Using default Python package layout
-- Found PY_em: /Library/Python/2.7/site-packages/em.pyc
-- Using empy: /Library/Python/2.7/site-packages/em.pyc
-- Using CATKIN_ENABLE_TESTING: ON
-- Call enable_testing()
-- Using CATKIN_TEST_RESULTS_DIR: /tmp/catkin_test_target_ws/build/testtest/test_results
-- Found gtest: gtests will be built
-- Using Python nosetests: /usr/local/bin/nosetests-2.7
-- catkin 0.6.9
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/catkin_test_target_ws/build/testtest
[testtest] <== '/private/tmp/catkin_test_target_ws/build/testtest/build_env.sh /usr/local/bin/cmake /private/tmp/catkin_test_target_ws/src/testtest -DCATKIN_DEVEL_PREFIX=/private/tmp/catkin_test_target_ws/devel -DCMAKE_INSTALL_PREFIX=/private/tmp/catkin_test_target_ws/install' finished with return code '0'


[testtest] ==> '/private/tmp/catkin_test_target_ws/build/testtest/build_env.sh /usr/bin/make test -j4 -l4' in '/private/tmp/catkin_test_target_ws/build/testtest'
Running tests...
Test project /tmp/catkin_test_target_ws/build/testtest
    Start 1: _ctest_testtest_rostest_test.test
1/1 Test #1: _ctest_testtest_rostest_test.test ...***Failed    1.16 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   1.16 sec

The following tests FAILED:
	  1 - _ctest_testtest_rostest_test.test (Failed)
Errors while running CTest
make: *** [test] Error 8
[testtest] <== '/private/tmp/catkin_test_target_ws/build/testtest/build_env.sh /usr/bin/make test -j4 -l4' failed with return code '2'

Failed   <== testtest [ 3.4 seconds ]
[build] There were '1' errors:

    Failed to build package 'testtest' because the following command:

        # Command run in directory: /private/tmp/catkin_test_target_ws/build/testtest
        /private/tmp/catkin_test_target_ws/build/testtest/build_env.sh /usr/bin/make test -j4 -l4

    Exited with return code: 2
Build summary:
 Failed     testtest
[build] Runtime: 3.5 seconds
```

- `catkin clean -a`
```
[clean] Removing buildspace: /private/tmp/catkin_test_target_ws/build
[clean] Removing develspace: /private/tmp/catkin_test_target_ws/devel
```

- `catkin_make_isolated --catkin-make-args test`
```
Base path: /private/tmp/catkin_test_target_ws
Source space: /private/tmp/catkin_test_target_ws/src
Build space: /private/tmp/catkin_test_target_ws/build_isolated
Devel space: /private/tmp/catkin_test_target_ws/devel_isolated
Install space: /private/tmp/catkin_test_target_ws/install_isolated
Additional make Arguments for catkin packages: test
~~
~~  traversing 1 packages in topological order:
~~  - testtest
~~
The packages or cmake arguments have changed, forcing cmake invocation

==> Processing catkin package: 'testtest'
==> Creating build directory: 'build_isolated/testtest'
==> cmake /private/tmp/catkin_test_target_ws/src/testtest -DCATKIN_DEVEL_PREFIX=/private/tmp/catkin_test_target_ws/devel_isolated/testtest -DCMAKE_INSTALL_PREFIX=/private/tmp/catkin_test_target_ws/install_isolated in '/private/tmp/catkin_test_target_ws/build_isolated/testtest'
-- The C compiler identification is AppleClang 6.0.0.6000056
-- The CXX compiler identification is AppleClang 6.0.0.6000056
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Using CATKIN_DEVEL_PREFIX: /tmp/catkin_test_target_ws/devel_isolated/testtest
-- Using CMAKE_PREFIX_PATH: /Users/william/indigo/install
-- This workspace overlays: /Users/william/indigo/install
-- Found PythonInterp: /usr/local/bin/python (found version "2.7.9")
-- Using PYTHON_EXECUTABLE: /usr/local/bin/python
-- Using default Python package layout
-- Found PY_em: /Library/Python/2.7/site-packages/em.pyc
-- Using empy: /Library/Python/2.7/site-packages/em.pyc
-- Using CATKIN_ENABLE_TESTING: ON
-- Call enable_testing()
-- Using CATKIN_TEST_RESULTS_DIR: /tmp/catkin_test_target_ws/build_isolated/testtest/test_results
-- Found gtest: gtests will be built
-- Using Python nosetests: /usr/local/bin/nosetests-2.7
-- catkin 0.6.9
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/catkin_test_target_ws/build_isolated/testtest
==> make test -j4 -l4 in '/private/tmp/catkin_test_target_ws/build_isolated/testtest'
Running tests...
Test project /tmp/catkin_test_target_ws/build_isolated/testtest
    Start 1: _ctest_testtest_rostest_test.test
1/1 Test #1: _ctest_testtest_rostest_test.test ...***Failed    0.87 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.87 sec

The following tests FAILED:
	  1 - _ctest_testtest_rostest_test.test (Failed)
Errors while running CTest
make: *** [test] Error 8
<== Failed to process package 'testtest':
  Command 'make test -j4 -l4' returned non-zero exit status 2

Reproduce this error by running:
==> cd /private/tmp/catkin_test_target_ws/build_isolated/testtest && make test -j4 -l4

Command failed, exiting.
```

It fails in both cases for me.

Even `catkin_make test`:

```
% ls
src

% catkin_make test
Base path: /private/tmp/catkin_test_target_ws
Source space: /private/tmp/catkin_test_target_ws/src
Build space: /private/tmp/catkin_test_target_ws/build
Devel space: /private/tmp/catkin_test_target_ws/devel
Install space: /private/tmp/catkin_test_target_ws/install
Creating symlink "/private/tmp/catkin_test_target_ws/src/CMakeLists.txt" pointing to "/Users/william/indigo/install/share/catkin/cmake/toplevel.cmake"
####
#### Running command: "cmake /private/tmp/catkin_test_target_ws/src -DCATKIN_DEVEL_PREFIX=/private/tmp/catkin_test_target_ws/devel -DCMAKE_INSTALL_PREFIX=/private/tmp/catkin_test_target_ws/install" in "/private/tmp/catkin_test_target_ws/build"
####
-- The C compiler identification is AppleClang 6.0.0.6000056
-- The CXX compiler identification is AppleClang 6.0.0.6000056
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Using CATKIN_DEVEL_PREFIX: /tmp/catkin_test_target_ws/devel
-- Using CMAKE_PREFIX_PATH: /Users/william/indigo/install
-- This workspace overlays: /Users/william/indigo/install
-- Found PythonInterp: /usr/local/bin/python (found version "2.7.9")
-- Using PYTHON_EXECUTABLE: /usr/local/bin/python
-- Using default Python package layout
-- Found PY_em: /Library/Python/2.7/site-packages/em.pyc
-- Using empy: /Library/Python/2.7/site-packages/em.pyc
-- Using CATKIN_ENABLE_TESTING: ON
-- Call enable_testing()
-- Using CATKIN_TEST_RESULTS_DIR: /tmp/catkin_test_target_ws/build/test_results
-- Found gtest: gtests will be built
-- Using Python nosetests: /usr/local/bin/nosetests-2.7
-- catkin 0.6.9
-- BUILD_SHARED_LIBS is on
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- ~~  traversing 1 packages in topological order:
-- ~~  - testtest
-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- +++ processing catkin package: 'testtest'
-- ==> add_subdirectory(testtest)
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/catkin_test_target_ws/build
####
#### Running command: "make test -j4 -l4" in "/private/tmp/catkin_test_target_ws/build"
####
Running tests...
Test project /tmp/catkin_test_target_ws/build
    Start 1: _ctest_testtest_rostest_test.test
1/1 Test #1: _ctest_testtest_rostest_test.test ...***Failed    1.10 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   1.10 sec

The following tests FAILED:
	  1 - _ctest_testtest_rostest_test.test (Failed)
Errors while running CTest
make: *** [test] Error 8
Invoking "make" failed
```

**Comment by wjwwood** [comment]
Also, I forgot to mention that I'm using your gist for the test.

**Comment by wjwwood** [comment]
Hmm, I can't explain why that's happening. I'll try to give it a look soon, but I'm not convinced it is specific to `catkin_tools` at this point.

**Comment by k-okada** [comment]
Humm, today everything has passed ok (after I rebooted my computer) => https://travis-ci.org/k-okada/testtest/builds/48161506 even in catkin_make test

**Comment by wjwwood** [comment]
@k-okada Ok... what does restarting your computer have to do with travis passing?

**Comment by k-okada** [comment]
oh, no, travis has passed, and testing on my local machine was passed after
reboot.
so now, my question is why it is not passed on your computer?...

--
◉ Kei Okada

**Comment by wjwwood** [comment]
I dunno, my local setup was in `/tmp` so it got ditched when I rebooted this weekend. I'll try to swing back around and test this again later.

**Comment by wjwwood** [comment]
With your test package I can also not reproduce it. I'm going to close this.

**Comment by k-okada** [comment]
could you reopen this? I finally able to create test code that fails on `catkin test`
https://travis-ci.org/k-okada/test_cmake_rostest/builds/49419166
it seems something wrong if we run build in parallel,  
I have noticed that  `build_env.sh` does not hold correct path if we build in parallel, but it may not related. (see 6th test)

**Comment by jbohren** [comment]
@k-okada I just took a look at this. The problem appears to be the `CATKIN_TEST_RESULTS_DIR` environment variable, which is saved in the `setup_cached.sh` file. If that environment variable points to the wrong package, then the tests will fail.

I found the following:

* `catkin test -p1` **pass**
* `catkin build -p1; catkin test -p1` **pass**
* `catkin build -p4; catkin test -p1` **fail**
* `catkin build -p4; catkin test testN` **pass** *(`testN` is the package in `CATKIN_TEST_RESULTS_DIR`)*
* `catkin build -p4; catkin test -p1 --force-cmake` **pass**

So for now if we want to make sure that `catkin test` works properly, we should probably change the alias to include `-p1 --force-cmake`.

**Comment by wjwwood** [comment]
@jbohren Why do you think it has to do with ``?

AFAIK, that has been removed:

- https://groups.google.com/forum/#!topic/ros-sig-buildsystem/suBvWzoGODU
- https://github.com/ros/catkin/issues/719
- https://github.com/ros/catkin/pull/728

If #259 is proposed as a temporary fix, then what is the long term solution? It isn't obvious to me why adding `--force-cmake` every time you run the tests is appropriate, could you expand on that?

**Comment by jbohren** [comment]
> @jbohren Why do you think it has to do with ``? [sic]

`CATKIN_TEST_RESULTS_DIR` in `setup_cached.sh` is the only thing that's different in the build space when building with `-p1` vs `-p4`.

> AFAIK, that has been removed:
> ...

Yeah I just discovered, this bug only appears to affect pre-Indigo versions of Catkin. I can reproduce it reliably on 12.04 / Hydro but not on 14.04 / Indigo. Sure enough, on Indigo, this file doesn't exist.

@k-okada Are you on Hydro or Indigo?

> If #259 is proposed as a temporary fix, then what is the long term solution? It isn't obvious to me why adding `--force-cmake` every time you run the tests is appropriate, could you expand on that?

The long term solution would be similar to what was done in ros/catkin#719 and ros/catkin#728, haha. So the options are:

- Leave in the `--force-cmake` hack (I don't like this)
- Backport the fixes in Catkin from Indigo to Hydro (could be a can of worms)
- Document that the `catkin test` alias doesn't work under Hydro under troubleshooting documentation (path of least resistance)

**Comment by wjwwood** [comment]
Well back porting to hydro isn't an option because the hydro farm has been shutdown and we no longer release new versions in hydro.

If `test` was a verb and not an alias we could:

- check the catkin version (in the workspace or in the system)
- if the version is too long to have the fix we warn or error
- the user can avoid the warning/error by forcing cmake

**Comment by jbohren** [comment]
How about we do the following two things: 
 - we add a `06-catkin-test-results-nuke.sh` to `/etc/catkin/profile.d` which `unset`s the variables
 - we modify the build env to set those two variables appropriately when building a given package

**Comment by jbohren** [comment]
@k-okada Give this a shot: https://github.com/catkin/catkin_tools/pull/265

**Comment by wjwwood** [comment]
I'm gonna close this with #265 being merged, please comment here if the problem persists.

**Comment by wjwwood** [comment]
Sorry this fix won't be in master until https://github.com/catkin/catkin_tools/pull/249 is merged. Reopening until then.

**Comment by jbohren** [comment]
This is fixed now that #249 is merged.
